bee-tangle/src/tangle.rs

```diff
@@ -1,759 +1,76 @@
-// Copyright 2020 IOTA Stiftung
-//
-// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
-// the License. You may obtain a copy of the License at
-//
-//     http://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
-// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and limitations under the License.
+use crate::{vertex::Vertex, TransactionRef};
 
-//! Module that provides the [`Tangle`] struct.
+use bee_transaction::{BundledTransaction as Transaction, Hash as TransactionHash, TransactionVertex};
 
-use crate::{
-    milestone::MilestoneIndex,
-    vertex::{TransactionRef, Vertex},
-};
+use dashmap::{mapref::entry::Entry, DashMap};
 
-use bee_transaction::{BundledTransaction as Transaction, Hash, TransactionVertex};
-
-use std::{
-    collections::HashSet,
-    sync::atomic::{AtomicU32, Ordering},
-};
-
-use async_std::{
-    sync::{Arc, Barrier},
-    task::block_on,
-};
-use dashmap::{mapref::entry::Entry, DashMap, DashSet};
-use flume::Sender;
-
-/// A datastructure based on a directed acyclic graph (DAG).
-pub struct Tangle {
-    /// A map between each vertex and the hash of the transaction the respective vertex represents.
-    pub(crate) vertices: DashMap<Hash, Vertex>,
-
-    /// A map between the hash of a transaction and the hashes of its approvers.
-    pub(crate) approvers: DashMap<Hash, Vec<Hash>>,
-
-    /// A map between the milestone index and hash of the milestone transaction.
-    milestones: DashMap<MilestoneIndex, Hash>,
-
-    /// A set of hashes representing transactions deemed solid entry points.
-    solid_entry_points: DashSet<Hash>,
-
-    /// The sender side of a channel between the Tangle and the (gossip) solidifier.
-    solidifier_send: Sender<Option<Hash>>,
-
-    solid_milestone_index: AtomicU32,
-    snapshot_milestone_index: AtomicU32,
-    last_milestone_index: AtomicU32,
-
-    drop_barrier: Arc<Barrier>,
+// TODO: Should we even add Key and Value type parameters?
+pub struct Tangle<Meta> {
+    pub vertices: DashMap<TransactionHash, Vertex<Meta>>,
```

Comment:
I think I'd opt for a more generic name than `Meta`. It could be anything really.

Reply:
Anything? What about .... `Meta`? :trollface:

Reply:
What about `Data`?

Reply:
Or maybe even just `T`?

Reply:
Thinking about it. The term `Metadata` really seems to be the best term for this situation. I mean ... the type parameter can be just `T` (that's fine for me), but the whole API is built around a `Vertex` encapsulating a `Transaction` and some additional information about that transaction, hence the term `metadata`. Not sure what you'ld prefer here.

Reply:
I'd personally go for T. That's the most generic one and doesn't "lock" us to anything.

Reply:
Done.

Reply:
the API is still `get_metadata`, though, and I am having a hard time of thinking about a better name for this.

Reply:
For the record: Thibault prefers `get_data`, but we decided (outside of this thread) to stick to the current API for the time being.